docker-build/Dockerfile

```diff
@@ -0,0 +1,68 @@
+FROM        debian 
+
+ARG         gsd
+ENV         GITHUB_STABLE_DIFFUSION $gsd
+
+ARG         sdreq="requirements-linux-arm64.txt"
+ENV         SD_REQ $sdreq
+
+ARG         condaarch
+ENV         ARCH $condaarch
+
+WORKDIR     /
+
+# TODO: Optimize image size
+
+COPY        entrypoint.sh anaconda.sh .
+SHELL       ["/bin/bash", "-c"]
+
+# Update and apt 446 MB
+RUN         apt update && apt upgrade -y \
+            && apt install -y \
+            git \
+            pip \
+            python3 \
+            wget 
+
+# install Anaconda or Miniconda 610 MB
+RUN         bash anaconda.sh -b -u -p /anaconda && /anaconda/bin/conda init bash
```

Comment:
This is over-engineered, particularly for a container.

Use the container's OS package manager to update Python if needed (and pip; some distros don't include it with Python:roll_eyes:), then use pip to install everything. The **only** reasons to use a venv are to 1) protect the OS Python system-packages 2) have completely separated environments - neither of these is necessary in a container.

If you're concerned about stepping on system Python packages, use pip's `--user` flag for everything. The only gotcha I know of is that you *might* have to add the user's Python Scripts directory to $PATH.

Reply:
Good point. This is a first version to get it working on a container but it's worth taking care of those optimizations sooner rather than later.

Reply:
TL;DR with this comment and the above comment about using the Mac reqs file:  switch to pip-only, and use the **Linux** reqs file, and I'm happy with the rest.

Reply:
Re: req files
But then the Dockerfile will need to know which platform (linux/**amd**64 or linux/**arm**64) it will be running on to install the right dependency. If it receives the appropriate dependency list to begin with, it remains agnostic on where it will run.

Also, while I don't have an Intel (**amd**64) Mac to test I strongly suspect the requirements-mac file won't work there, confusing users. That use case (Linux container on Intel Mac) probably needs a linux-arm64 reqs file i.e. the current Linux reqs file.

Reply:
I'm still not getting the use-case here. To my mind, it's greatest potential is in a hosted environment, where

- one doesn't want to **even think** about using ARM for performance reasons
- ROCm *might* be an option except...
- most hosting services (which I'm aware of) use CUDA-capable hardware

AFA locally running SD in a container... why? It makes more sense in every way I can think of to simply set up a Python venv

Reply:
Yes, in production you won't be using arm64.
Locally:
- For regular use you'll want to take advantage of your GPUs by setting up a Python env.
- But locally for some development tasks (like testing changes to the Dockerfile as I'm doing right now) you'll want to have a Docker setup if you're testing changes specific to that container functionality.

Reply:
I don't think that local development and testing of the changes in the PR itself are sufficient reason to complicate the PR itself.

Keep a local stash of the stuff you need to do dev work, which you can apply when needed.

Reply:
> I don't think that local development and testing of the changes in the PR itself are sufficient reason to complicate the PR itself.
> 
> Keep a local stash of the stuff you need to do dev work, which you can apply when needed.

After Docker support is added wouldn't we want to make it easier for developers to contribute to that functionality? All it takes is giving them the right req list for what they're trying to do and they're up and running. Instead of having to figure out that they need a different fork of k-diffusion in their scenario.

Reply:
Right, but we already have a massive proliferation of `requirements-*.txt` files, and `environment` files, and installation methods - it ends up being VERY confusing for new users.

I already have a PR open to vastly simplify the requirements.txt landscape. The one I'll open after that one will completely drop conda. and substitute pure-Python 'pew' and 'pip'

(I'm being a bit silly here, but...)

> wouldn't we want to make it easier for developers to contribute to that functionality

No. Containerizing this repo is not the main focus here. If there's any QOL improvements for devs to be made, it's going to be in docs, refactoring, PR actions to automatically catch simple errors, that sort of thing.

I hope I'm not coming off as hostile: I think you've done good work, and it's a cool idea, but this-repo-as-a-hosted-microservice is just not in the near-term for most users or devs.

One idea which I had would be for you to create a "companion" repo (NOT a fork of this repo - I don't think that's necessary) which could house all of the work necessary to do what you're trying to do (for example: if needed, you could `git clone` this repo into the image at build time). At some later date, we could then possibly merge the two (which **shouldn't** be that hard, as generally speaking you wouldn't be touching the internals in *this* repo)

Reply:
I think we're getting to the same goal here, adding a deployment option for Docker so it's not limited to people's laptops (though it's true that's the most common scenario). And I like the idea of dropping conda completely (this PR will do that for containers once your suggestion is implemented. That was very helpful, thanks).

I think we can move directly into that future state you mention (two repos merged) by making sure deployment options (macOS, Windows, Linux; amd64 vs arm64) remain isolated from project internals in this PR. Being a fork is crucial, though. That way I can sync changes that affect the Dockerfile like the change in how GFPGAN/Real-ESRGAN is installed.  

I can add the instructions for modifying your requirements file for the different scenarios. This makes sense now that you mention the whole requirements management will be overhauled in the near future. How does that sound?

Reply:
Sounds good